ENH add generalized inverse Gaussian distribution to scipy.stats

chrisb83 opened the pull request:
This is a well-known distribution. Sampling of random  variates is non-trivial. Approach here relies on ratio of uniforms method (#8293) and a recent algorithm of Hoermann/ Leydold.

PR 8293 is still under review, so I added it in stats/stats.py this branch as well for testing (can be removed later easily).

Tutorial file is still WIP.

chrisb83 commented:
`FAIL scipy/stats/tests/test_continuous_basic.py::test_rvs_broadcast[geninvgauss-shape_args35]`

At the moment, I don't see an easy way to extend the rvs method to work with non-scalar parameters of the distribution. For given parameters `p, b`, an iterative procedure is used to generate the random variates. Is there  a simple way to ensure that broadcasting works if `_rvs` already works for scalars?

WarrenWeckesser commented:
Probably the simplest way is to write the scalar version, and then create a vectorized wrapped using `numpy.vectorize`.

chrisb83 commented:
Hi Warren,

I am struggling with the correct use of `np.vectorize` here...
- should I try to avoid to use `np.vectorize` if the input is scalar or is it not worth it?
- if I specify `otypes=[np.ndarray]`, the output is an array of `dtype=object` that has arrays as elements (causing some tests to fail). is this there a better way? or how should the results be transformed? (see example below)

Maybe you can help ...

I tried the following code (not committed):
```
def _rvs(self, p, b):
        if (np.isscalar(p) and np.isscalar(b)) or (p.size == 1 and b.size == 1):
            return self._rvs_scalar(np.asscalar(p), np.asscalar(b))
        else:
            _rvs_vec = np.vectorize(self._rvs_scalar, otypes=[np.ndarray])
            return _rvs_vec(p, b)
```
`geninvgauss([1, 2, 3], 1).rvs()` returns 
```
array([array([0.49563439, 1.81470995, 0.68474693]),
       array([3.34586756, 0.9972639 , 2.94646317]),
       array([7.84683634, 7.1318319 , 5.10001549])], dtype=object)
```

chrisb83 commented:
Thanks, I hope there is a simple solution :) I think one needs to take into account that `size` and the shape parameters at the same time, see the example below.

`test` mimics my method `_rvs_scalar` (although test can handle an array `p`). 
```
def test(size, p):
    return stats.norm.rvs(size=size, scale=p)

test_vec = np.vectorize(test, otypes=[np.ndarray], excluded='size')

test_vec(size=3, p=[1, 2, 3])
array([array([-1.36900502,  0.64463791, -2.56062163]),
       array([-1.63466252, -0.49033627, -1.88337739]),
       array([-0.64243163, -1.58856722,  6.09658811])], dtype=object)

test(size=3, p=[1, 2, 3])
array([-0.29158522, -1.18783153,  4.58038196])
```

Or if `size=4`, the vectorized version returns a results while `test(...)`.

WarrenWeckesser commented:
Chris, sorry to mislead you about using `vectorize`.  It doesn't really help in your case.

Here's one way to handle the case where you have a function that generates samples for a fixed set of scalar shape parameters.  I can't guarantee that it is the simplest approach, but it works as far as I've tested it.  In the following, I created an example by subclassing the `f_gen` distribution, and giving it the method `_rvs_scalar()` that accepts scalar shape parameters and an integer number of variates to generate.  (I made it a method, but it could just as well be an independent function--it does not use `self`.)  The `_rvs()` method of `fx_gen` uses `_rvs_scalar()` to generate its result.

The `"__main__"` script at the end shows an example where broadcasting is used to generate samples for four different sets of shape parameters with one call to `rvs()`.

```
from itertools import zip_longest
import numpy as np
from scipy.stats.distributions import f_gen
import matplotlib.pyplot as plt


def _check_shape(argshape, size):
    """
    This is a utility function used by `_rvs()` in the class fx_gen
    defined below.  It compares the tuple argshape to the tuple size.

    The function returns two tuples, scalar_shape and bc.

    scalar_shape is the shape to which the 1-d array of random variates
    returned by _rvs_scalar() is converted when it is copied into the
    output array of _rvs().

    bc is an tuple the same length as size that contains bools.
    bc[j] is True if the data associated with that index is generated
    in one call of _rvs_scalar().
    """
    scalar_shape = []
    bc = []
    for argdim, sizedim in zip_longest(argshape[::-1], size[::-1], fillvalue=1):
        if sizedim > argdim or (argdim == sizedim == 1):
            scalar_shape.append(sizedim)
            bc.append(True)
        else:
            bc.append(False)
    return tuple(scalar_shape[::-1]), tuple(bc[::-1])


class fx_gen(f_gen):

    def _rvs_scalar(self, dfn, dfd, numsamples):
        # For *scalars* dfn and dfd (the shape parameters
        # of this distribution), generate `numsamples` random variates.
        # The location and scale parameters of the distribution must be
        # taken to be 0 and 1, respectively.
        # `numsamples` will be a nonnegative integer; it can be 0.
        sample = ((np.random.chisquare(dfn, size=numsamples) / dfn) /
                  (np.random.chisquare(dfd, size=numsamples) / dfd))
        return sample

    def _rvs(self, dfn, dfd):
        # When this method is called, self._size will be a (possibly empty)
        # tuple of integers.  It will not be None; if `size=None` is passed
        # to `rvs()`, self._size will be the empty tuple ().

        dfn, dfd = np.broadcast_arrays(dfn, dfd)
        # dfn and dfd now have the same shape.

        # `shp` is the shape of the blocks of random variates that are
        # generated for each combination of parameters associated with
        # broadcasting dfn and dfd.
        # bc is a tuple the same lenth as self._size.  The values
        # in bc are bools.  If bc[j] is True, it means that
        # entire axis is filled in for a given combination of the broadcast
        # arguments.
        shp, bc = _check_shape(dfn.shape, self._size) 

        #`numsamples` is the total number of variates to be generated
        # for each combination of the input arguments.
        numsamples = int(np.prod(shp))

        # `out` is the array to be returned.  It is filled in in the
        # loop below.
        out = np.empty(self._size)

        it = np.nditer([dfn, dfd],
                       flags=['multi_index'],
                       op_flags=[['readonly'], ['readonly']])
        while not it.finished:
            # Convert the iterator's multi_index into an index into the `out`
            # array where the call to _rvs_scalar() will be stored.
            # Where bc is True, we use a full slice; otherwise we use the
            # index value from it.multi_index.  len(it.multi_index) might be
            # less than len(bc), and in that case we want to align these
            # two sequences to the right, so the loop variable j runs from
            # -len(self._size) to 0.  This doesn't cause an IndexError, because
            # bc[j] will be True in those cases where it.multi_index[j]
            # would cause an IndexError. 
            idx = tuple((it.multi_index[j] if not bc[j] else slice(None))
                         for j in range(-len(self._size), 0))
            out[idx] = self._rvs_scalar(it[0], it[1],
                                        numsamples).reshape(shp)
            it.iternext()

        if self._size == ():
            # This is not essential.  Without it, though, a call to
            # `rvs(..., size=None)`, which generates a single sample,
            # returns the scalar array `np.array(value)` instead of
            # just `value`.  We only do this for _size == (), not for
            # _size == (1,); the latter means an array with length 1,
            # i.e. `np.array([value])`.
            out = out[()]

        return out



if __name__ == "__main__":
    # This demo creates four plots.  The data in the histograms is
    # generated with one call to `fx.rvs()` using broadcasting.
    # The PDF curves are computed individually.  (These also could be
    # computed using broadcasting.)

    fx = fx_gen()
    dfn = np.array([[10], [100]])
    dfd = np.array([10, 100])
    s = fx.rvs(dfn, dfd, size=(40000, 2, 2))

    x = np.linspace(1e-8, 5, 400)
    for i in range(2):
        dfnum = dfn[i, 0]
        for j in range(2):
            dfden = dfd[j]
            plt.subplot(2, 2, 1+ 2*i + j)
            sample = s[:, i, j]
            plt.hist(sample, bins=250, alpha=0.5, normed=True)
            plt.plot(x, fx.pdf(x, dfnum, dfden))
            plt.xlim(0, 5)
            plt.title('dfn=%d, dfd=%d' % (dfnum, dfden))
    plt.tight_layout()
    plt.show()
```

Here's the plot generated by the script:

![figure_1](https://user-images.githubusercontent.com/321463/39055372-103376fe-4482-11e8-91af-b0068c355255.png)

chrisb83 commented:
Thanks for your help, much appreciated. I think it should work now. 

I added _check_shape to _distn_infrastructure. Is that the right place?

I still need to resove the conflicts compared to the master. From the development guide, I understand that rebasing is the preferred way of doing this (https://docs.scipy.org/doc/numpy/dev/gitwash/development_workflow.html)?

chrisb83 commented:
Closes #6537

chrisb83 commented:
@WarrenWeckesser Hi Warren, now that the ratio of uniforms method has been merged, I want to continue to work on this PR. Since you were involved in the review of RoU, could you take a look here as well if you have time? Travis-CI passes, CircleCI crashed while runnning tests related geninvgauss (could be a coincidence, to be checked if it runs again).

chrisb83 commented:
@WarrenWeckesser Hi Warren, do you have time to look at this PR any time soon? Thx

WarrenWeckesser commented:
@chrisb83, I'll get back to it this weekend.

WarrenWeckesser commented:
I'll have more comments tomorrow, but for now here's a small problem that I noticed.  The `pdf` method generates a warning and returns `nan` when evaluated at `x=0`:

```
In [44]: geninvgauss.pdf(0, 0.5, 2)                                             
/Users/warren/miniconda36/lib/python3.6/site-packages/scipy-1.3.0.dev0+6cc598a-py3.6-macosx-10.7-x86_64.egg/scipy/stats/_continuous_distns.py:3355: RuntimeWarning: divide by zero encountered in power
  return np.power(x, p-1) * np.exp(-b*(x+1/x)/2) / (2*sc.kv(p, b))
/Users/warren/miniconda36/lib/python3.6/site-packages/scipy-1.3.0.dev0+6cc598a-py3.6-macosx-10.7-x86_64.egg/scipy/stats/_continuous_distns.py:3355: RuntimeWarning: divide by zero encountered in true_divide
  return np.power(x, p-1) * np.exp(-b*(x+1/x)/2) / (2*sc.kv(p, b))
/Users/warren/miniconda36/lib/python3.6/site-packages/scipy-1.3.0.dev0+6cc598a-py3.6-macosx-10.7-x86_64.egg/scipy/stats/_continuous_distns.py:3355: RuntimeWarning: invalid value encountered in multiply
  return np.power(x, p-1) * np.exp(-b*(x+1/x)/2) / (2*sc.kv(p, b))
Out[44]: nan
```
`x=0` will have to be handled separately to avoid the problem caused by the term `1/x` in the formula for the PDF.

chrisb83 commented:
> x=0 will have to be handled separately to avoid the problem caused by the term 1/x in the formula for the PDF.

You are right, unfortunately Python does not understand that the term converges :)

WarrenWeckesser commented:
@chrisb83, I'm currently looking into the accuracy of the default CDF and SF methods for this distribution.  I'm checking with Wolfram Alpha's [`InverseGaussianDistribution`](https://reference.wolfram.com/language/ref/InverseGaussianDistribution.html`), and with `mpmath`.  For the `mpmath` code, I'll add a few of relevant functions in a `geninvgauss` distribution in my [`mpsci`](https://github.com/WarrenWeckesser/mpsci) repository.

A change that should definitely be made before merging is to implement the `_logpdf` method explicitly.  I haven't experimented with the `fit` method yet, but having the explicit `_logpdf` should at least make `fit` a bit more efficient.

WarrenWeckesser commented:
When calling the `rvs` method with moderately large values of `p`, I occasionally got some warnings.  I tracked down the source of the warnings to the following problem that occurs in `_pdf` when `x` is large and `p` is moderately large:

```
In [137]: geninvgauss.pdf(2e6, 50, 2)                                           
/Users/warren/miniconda36/lib/python3.6/site-packages/scipy-1.3.0.dev0+6cc598a-py3.6-macosx-10.7-x86_64.egg/scipy/stats/_continuous_distns.py:3355: RuntimeWarning: overflow encountered in power
  return np.power(x, p-1) * np.exp(-b*(x+1/x)/2) / (2*sc.kv(p, b))
/Users/warren/miniconda36/lib/python3.6/site-packages/scipy-1.3.0.dev0+6cc598a-py3.6-macosx-10.7-x86_64.egg/scipy/stats/_continuous_distns.py:3355: RuntimeWarning: invalid value encountered in multiply
  return np.power(x, p-1) * np.exp(-b*(x+1/x)/2) / (2*sc.kv(p, b))
Out[137]: nan
```

The term `power(x, p-1)` overflows.  The PDF should evaluate to 0.0, because the exponential term will be small enough to squash the power term.

I think the problem can be avoided if we implement the `_logpdf` method and then implement `_pdf(...)` as `exp(self._logpdf(...))`.

chrisb83 commented:
Thanks for pointing this out.

> I think the problem can be avoided if we implement the _logpdf method and then implement _pdf(...) as exp(self._logpdf(...)).

Indeed, logpdf was already implemented, I use it now in pdf (np.exp(...)) as suggested. 

I need to rebase though...

chrisb83 commented:
I added testcases for `pdf(0)` and `geninvgauss.pdf(2e6, 50, 2)`, that works now.

CircleCI crashed again when running tests related to rvs sampling. I think the check relying on the KS-test is very slow since it relies on the cdf which need to be computed numerically. I simplified some of the tests to save time. Still the generic test of rvs failed involving the standard parameters in `distr_params.py`. I will check this in more detail

chrisb83 commented:
I checked the default test on `rvs` that uses `N=1000` and the parameters 2.3, 1.5 for `geninvgauss`.
The test takes about 47 seconds on my machine and the p-values do not look suspicious (I did 10 repetitions).

```
pval = []
t = []
np.random.seed(2356)
for i in range(10):
    rvs = geninvgauss(2.3, 1.5).rvs(size=1000)
    tic = time.time()
    pval.append(stats.kstest(rvs, geninvgauss(2.3, 1.5).cdf)[1])
    t.append(time.time() - tic)

pval: [0.732882800373323,
 0.8152476879864183,
 0.17837665493242233,
 0.7768452212121085,
 0.9769002909456006,
 0.5432102182406183,
 0.10360341632109987,
 0.49116284749357875,
 0.6681503209560733,
 0.7520084378567328]

t:
[47.51063370704651,
 46.95862436294556,
 47.72902846336365,
 47.05382513999939,
 47.250492811203,
 47.36603260040283,
 47.28733682632446,
 47.47200036048889,
 53.51115274429321,
 50.062657833099365]
```

So I don't think there is a problem with the implementation. Any ideas how to proceed to avoid the test failures?

WarrenWeckesser commented:
@chrisb83, I'm getting back to this PR.  Let's see if we can get it into 1.3.

The tests of `rvs` are quite slow.  It is probably worth spending some time trying to improve the performance, but I don't have any concrete suggestions for that at the moment.

One thing that needs to be fixed is when a scalar sample is requested (i.e. `size` is not specified, or `size=None`).  A scalar should be returned, but currently it returns a 1-d array with length 1:

    In [12]: geninvgauss.rvs(4, 3, 2, 1)                                                                                                                                                     
    Out[12]: array([5.15516614])

Compare that to, for example, `gamma.rvs()`:

    In [13]: gamma.rvs(3, 2, 1)                                                                                                                                                              
    Out[13]: 7.020025223999881

The return value *should* be a 1-d array when `size=1` is given (and it is):

    In [29]: geninvgauss.rvs(4, 3, 2, 1, size=1)                                                                                                                                             
    Out[29]: array([4.9732858])

chrisb83 commented:
Sure, I will check why the output is always an array

Regarding performance: the problem is the calculation of the cdf, the rv generation is fast. I don't think this can be fixed easily since we would need a good approximation method. Let me see if I can find a paper on this.

chrisb83 commented:
The following R package mentions a paper on efficient integration:
https://cran.r-project.org/web/packages/GeneralizedHyperbolic/GeneralizedHyperbolic.pdf

namely:

https://www.sciencedirect.com/science/article/pii/S0168927410000693 (I don't have access)

I suspect this would be a lot of work :)

tylerjereddy commented:
@chrisb83 I emailed you a PDF of the full article

chrisb83 commented:
I had a look at the paper which indeed proposes an approximation of the relevant integral: it is called incomplete Bessel function and it is related to the generalized incomplete Gamma function, which are not in `scipy.special` to my knowledge.

As expected, this is likely to be quite a bit of work and I would prefer if we leave such an improvement to a future project, I don't have experience in this area.

Would skipping the usual ks-test that checks the distributions be an option? I could add the relevant test to `test_rvs` for the same distribution which is already marked as slow.

WarrenWeckesser commented:
@chrisb83 wrote

> As expected, this is likely to be quite a bit of work and I would prefer if we leave such an improvement to a future project

That's fine with me.  We can leave as a "wish list" item.

WarrenWeckesser commented on scipy/stats/_continuous_distns.py:
```diff
@@ -3338,6 +3338,220 @@ def _stats(self, mu):
+        elif b >= min(0.5, 2 * np.sqrt(1 - p) / 3):
+            # ratio of uniforms without mode shift
+            umax = np.sqrt(self._pdf(m, p, b))
+            s = sc.kv(p + 2, b) / sc.kv(p, b)
```
You have 

            s = sc.kv(p + 2, b) / sc.kv(p, b)
            vmax = np.sqrt(self._pdf(self._mode(p + 2, b), p + 2, b) * s)

The version in the paper, translated to the code's variable names, would be

            xplus = ((1 + p) + sqrt((1 + p)**2 + b**2))/b
            vmax = xplus * np.sqrt(self._pdf(xplus, p, b))

After working through the algebra I see that mathematically they are the same, but I don't understand why you've expressed the calculation the way you did.  It looks like a confusing way to express the formula, and it also seems like it would be computationally more expensive--there are two additional calls to `kv` that the straight-forward version doesn't have.  Is there some advantage to your version that I don't see?

chrisb83 replied:
no, i think it is better to use the version stated in the paper, fixed

WarrenWeckesser commented on scipy/stats/_continuous_distns.py:
```diff
@@ -3338,6 +3338,220 @@ def _stats(self, mu):
+            N = np.prod(size1d)  # number of rvs needed, reshape upon return
+
+            # start sampling
+            rng = self._random_state
```
Similar to the comment above about `rndm`:  `rng` is only used in two places below, so I don't see the advantage that creating this name provides.

chrisb83 replied:
fixed

WarrenWeckesser commented on scipy/stats/_continuous_distns.py:
```diff
@@ -3338,6 +3338,220 @@ def _stats(self, mu):
+            x = np.zeros(N)
+            simulated = 0
+            # [2]: rejection constant is < 2.73; so expected runtime is finite
+            while True:
```
This loop begins with `while True:`, and is terminated with a `break` in an `if` statement:

                if simulated >= N:
                    rvs = np.reshape(x, size1d)
                    break

You have constructed the calculation so that `simulated` *can't* exceed `N`, and the loop only terminates when `simulated == N`.  So it would be cleaner to start the loop with

            while simulated < N:

then remove the `if` statement that checks for the termination condition, and after the while loop add the line

            rvs = np.reshape(x, size1d)

chrisb83 replied:
you are right, fixed

WarrenWeckesser commented on scipy/stats/_continuous_distns.py:
```diff
@@ -3338,6 +3338,220 @@ def _stats(self, mu):
+                k = N - simulated
+                h, rvs = np.zeros(k), np.zeros(k)
+                # simulate uniform rvs on [x1, x2] and [0, y2]
+                u1 = rng.random_sample(size=k)
```
Style nit:  you use the variable names `u1` and `v1`, which leads me to expect there will be other variables such as `u2` and `v2`, or `u0` and `v0`.  But there are no other variables like that, so these variables should just be `u` and `v`.  (That also matches the paper, where there are no numerical subscripts on the corresponding variables U and V.)

chrisb83 replied:
fixed

WarrenWeckesser commented on scipy/stats/_continuous_distns.py:
```diff
@@ -3338,6 +3338,220 @@ def _stats(self, mu):
+                # apply rejection method
+                accept = (u1 * h <= self.pdf(rvs, p, b))
+                num_accept = sum(accept)
+                if num_accept > 0:
```
By construction, `num_accept` can't be more than `N - simulated`, so these statements

                if num_accept > 0:
                    take = min(num_accept, N - simulated)
                    x[simulated:(simulated + take)] = rvs[accept][0:take]
                    simulated += take

can be simplified to

                if num_accept > 0:
                    x[simulated:(simulated + num_accept)] = rvs[accept]
                    simulated += num_accept

chrisb83 replied:
good observation; fixed

WarrenWeckesser commented on scipy/stats/_continuous_distns.py:
```diff
@@ -3338,6 +3338,220 @@ def _stats(self, mu):
+        return rvs
+
+    def _mode(self, p, b):
+        return b / (np.sqrt((p - 1)**2 + b**2) + 1 - p)
```
This is the formula appropriate for p < 1, but it has problems when p > 1 and b is small.  The problem shows up in a call such as `geninvgauss.rvs(1.5, 1e-8)`, which generates the warnings shown below and fails to generate samples.

The problem is that when p > 1, and b is small, the formula for the mode suffers from catastrophic loss of precision.  It looks like this can be fixed by selecting the formula for the mode (equation (2) in the paper)
based on p's size relative to 1:

    def _mode(self, p, b):
        if p < 1:
            m = b / (np.sqrt((p - 1)**2 + b**2) + 1 - p)
        else:
            m = (np.sqrt((1 - p)**2 + b**2) - (1 - p)) / b
        return m

Here's the result of calling `geninvgauss.rvs(1.5, 1e-8)`:

```
In [34]: geninvgauss.rvs(1.5, 1e-8)                                                                          
/Users/warren/miniconda36/lib/python3.6/site-packages/scipy-1.3.0.dev0_-py3.6-macosx-10.7-x86_64.egg/scipy/stats/_continuous_distns.py:3546: RuntimeWarning: divide by zero encountered in double_scalars
  return b / (np.sqrt((p - 1)**2 + b**2) + 1 - p)
/Users/warren/miniconda36/lib/python3.6/site-packages/scipy-1.3.0.dev0_-py3.6-macosx-10.7-x86_64.egg/scipy/stats/_continuous_distns.py:3461: RuntimeWarning: invalid value encountered in double_scalars
  p1 = a1 - a2**2 / 3
/Users/warren/miniconda36/lib/python3.6/site-packages/scipy-1.3.0.dev0_-py3.6-macosx-10.7-x86_64.egg/scipy/stats/_continuous_distns.py:3462: RuntimeWarning: invalid value encountered in double_scalars
  q1 = 2 * a2**3 / 27 - a2 * a1 / 3 + m
/Users/warren/miniconda36/lib/python3.6/site-packages/scipy-1.3.0.dev0_-py3.6-macosx-10.7-x86_64.egg/scipy/stats/_continuous_distns.py:3384: RuntimeWarning: invalid value encountered in subtract
  lambda x, p, b, c: c + (p-1)*np.log(x) - b*(x+1/x)/2,
---------------------------------------------------------------------------
RuntimeError                              Traceback (most recent call last)
<ipython-input-34-6c4504cd3606> in <module>
----> 1 geninvgauss.rvs(1.5, 1e-8)

~/miniconda36/lib/python3.6/site-packages/scipy-1.3.0.dev0_-py3.6-macosx-10.7-x86_64.egg/scipy/stats/_distn_infrastructure.py in rvs(self, *args, **kwds)
    951         # by _rvs().
    952         self._size = size
--> 953         vals = self._rvs(*args)
    954 
    955         vals = vals * scale + loc

~/miniconda36/lib/python3.6/site-packages/scipy-1.3.0.dev0_-py3.6-macosx-10.7-x86_64.egg/scipy/stats/_continuous_distns.py in _rvs(self, p, b)
   3395             out = self._rvs_scalar(p, b, self._size)
   3396         elif p.size == 1 and b.size == 1:
-> 3397             out = self._rvs_scalar(p.item(), b.item(), self._size)
   3398         else:
   3399             # When this method is called, self._size will be a (possibly empty)

~/miniconda36/lib/python3.6/site-packages/scipy-1.3.0.dev0_-py3.6-macosx-10.7-x86_64.egg/scipy/stats/_continuous_distns.py in _rvs_scalar(self, p, b, numsamples)
   3471             rvs = rvs_ratio_uniforms(lambda x: self.pdf(x, p, b), size=sz,
   3472                                      umax=umax, vmin=vmin, vmax=vmax, c=m,
-> 3473                                      random_state=rndm)
   3474         elif b >= min(0.5, 2 * np.sqrt(1 - p) / 3):
   3475             # ratio of uniforms without mode shift

~/miniconda36/lib/python3.6/site-packages/scipy-1.3.0.dev0_-py3.6-macosx-10.7-x86_64.egg/scipy/stats/_rvs_sampling.py in rvs_ratio_uniforms(pdf, umax, vmin, vmax, size, c, random_state)
    166                    "to work for the provided parameters. Please check the "
    167                    "pdf and the bounds.".format(i*N))
--> 168             raise RuntimeError(msg)
    169         i += 1

RuntimeError: Not a single random variate could be generated in 50000 attempts. The ratio of uniforms method does not appear to work for the provided parameters. Please check the pdf and the bounds.
```

chrisb83 replied:
you are right, i incorrectly assumed the version I had works for both cases, fixed now

chrisb83 replied:
and it checked it: `geninvgauss.rvs(1.5, 1e-8)` returns a value immediately in this extreme case
`geninvgauss(1.5, 1e-8).stats() # (array(3.e+08), array(6.e+16))

chrisb83 commented on scipy/stats/_continuous_distns.py:
```diff
@@ -3338,6 +3338,220 @@ def _stats(self, mu):
+            root1 = s1 * np.cos(phi / 3 + np.pi / 3) - a2 / 3
+            root2 = -s1 * np.cos(phi / 3) - a2 / 3
+            # root3 = s1 * np.cos(phi / 3 - np.pi / 3) - a2 / 3
+            vmin = (root1 - m) * np.sqrt(self.pdf(root1, p, b))
```
fixed

chrisb83 commented on scipy/stats/_continuous_distns.py:
```diff
@@ -3338,6 +3338,220 @@ def _stats(self, mu):
+            p = -p
+            invert_res = True
+        m = self._mode(p, b)
+        if (p >= 1) | (b > 1):
```
fixed

chrisb83 commented on scipy/stats/_continuous_distns.py:
```diff
@@ -3338,6 +3338,220 @@ def _stats(self, mu):
+
+    def _rvs_scalar(self, p, b, numsamples=None):
+        invert_res = False
+        sz, rndm = numsamples, self._random_state
```
fixed

chrisb83 commented on scipy/stats/_continuous_distns.py:
```diff
@@ -3338,6 +3338,220 @@ def _stats(self, mu):
+
+    References
+    ----------
+    .. [1] O. Barndorff-Nielsen, P. Blaesild, C. Halgreen, "First hitting time
```
i added [1]_ after the cdf

chrisb83 commented:
thanks Warren for the thorough review. Some of your observations can also be used to improve `rvs_ratio_uniforms` in another PR.

rvs should now return a scalar if size is None. This seems to be an issue of rvs_ratio_uniforms as well.

I tried to skip the kstest in tests/test_continuous_basic.py and added such tests of rvs as specific tests that are marked as slow.

I still need to investigate two things:

- shape / broadcasting issue you mentioned
- extreme parameters when the method is very slow (this should not be the case)

(Probably I won't have much time before 1.3.0, so maybe this PR has to wait)

chrisb83 commented:
Hi Warren, 

the issue of the `pdf` is related to my change to get the right value at zero using `_lazywhere` in `_logpdf`. 
If I simply write `np.where(x > 0, c + (p-1)*np.log(x) - b*(x+1/x)/2, -np.inf)`, I get the expected result with shape (2, 5) but this raises a runtime warning since the function is evaluated at zero although the condition is false. Do you have a simple solution for that?

chrisb83 commented:
I started to look into the issue related to rvs for the extreme parameters:

**Ratio of uniforms with mode shift:**
p, b = 250, 1.8: root1 and root2 are large such that pdf = 0 => vmin = vmax == 0 and rv_ratio_uniforms raises an error
p, b = 2.5, 1000: now sc.kv(p, b) is zero when evaluating the logpdf

**New algorithm**
p, b = 0.75, 1e-6: need to invetigate further, variables xs, A1, A2 etc computed in the method look ok

tylerjereddy commented:
Sounds like this may not be quite ready for the branching of 1.3.x in 6 days? I won't bump the milestone just yet but I do see:

> Probably I won't have much time before 1.3.0, so maybe this PR has to wait

WarrenWeckesser commented:
I'll get back to it this weekend.  Maybe we can get the remaining issues resolved before 1.3 branches.

WarrenWeckesser commented:
After far too many false starts, I realized that the problem with the broadcasting in the `pdf`, `logpdf` and other methods could be fixed by implementing `_argcheck(self, p, b)` as:

        def _argcheck(self, p, b):
            return (p == p) & (b > 0)

That ensures that the result of `_argcheck` broadcasts the `p` and `b` arguments.

chrisb83 commented:
Thanks, I was not aware of that, so that is easy to fix.

I think the main functionality should be ok after the fix of _argcheck at least for the most common use cases and I won't be able to fix these issues relates to extreme parameters before 1.3.0. 

Are your other concerns comments addressed apart from the extreme parameters?

WarrenWeckesser commented:
@chrisb83, I found the cause of the poor performance when b is small.  The problem is that bounding functions s2 and h2 (and the corresponding constant A2) described in Lemma 14 are based specifically on the function g(x) defined in equation (1).  So we can't use the PDF to evaluate the constant k1 or to check for acceptance of the random variate.  The specific form of g(x) shown in the paper must be used.

When I changed the code to use g(x), the number of iterations required to accept a random variate in this case:

    np.random.seed(123)
    x = geninvgauss.rvs(0.75, 1e-6)

went down from almost 300000 to 1!  The unit tests still passed, and a few more nonrigorous checks in which I visually compared the histogram to the PDF didn't show any problems.

I'll spend some time today checking if this affects the other methods.

chrisb83 commented:
> The problem is that bounding functions s2 and h2 (and the corresponding constant A2) described in Lemma 14 are based specifically on the function g(x) defined in equation (1). 

Indeed, you are right, i overlooked this, so THANKS for spotting this mistake. I always assumed g is the density. I think this affects rvs even in case the ratio of uniforms is used since v+ ans v- depend in g. Nothing difficult but I won't have time before 1.3 to work in it. One needs to be careful with the notation, even the mode in the paper is w.r.t. to g, not the pdf.

chrisb83 commented:
@tylerjereddy i removed the 1.3.0 Milestone, this PR needs more work

tylerjereddy commented:
@chrisb83 Thanks for letting me know

pvanmulbregt commented on doc/source/tutorial/stats/continuous_geninvgauss.rst:
```diff
@@ -0,0 +1,20 @@
+	        f(x; p, b) = x^{p-1} \exp(-b(x + 1/x)/2) / (2 K_p(b)),
+	\end{eqnarray*}
+
+where :math:`x > 0` is a real number and the parameters :math:`p, b` satisfy :math:`b > 0`. :math:`K_v` is the modified Bessel function of second kind of order :math:`p` (`scipy.special.kv`).
```
Should K_v  be K_p, as in ":math:`K_p`"?

chrisb83 replied:
yes, i changed :math:`p` to :math:`v` to have better alignment to `Special.kv`

WarrenWeckesser commented:
@chrisb83, OK, we can get this into 1.4 for sure.

chrisb83 commented:
Hi Warren,

after your feedack, I had another look at the paper and changed the implementation: I added the quasi-density and use it for the rv generation.

A few comments:
- for `_mode`, it does not matter since the quasi-density only rescaled by a constant
- in the ratio of uniforms method, one could also use the density (as before) since the constants simply cancel out. however, using the quasidensity avoids the calculation of the constant (`special.kv(...)`)
- The new algorithm based on finding a function h that is larger than g is indeed adapted to the function without the normalizing factor. after the code change, both `p, b = 0.75, 1e-6` and `p, b = 0.75, 1e-8` return immediately (I tried `size=1000`)

Returning to the problematic cases above, the status now is:

p, b = 250, 1.8: same as above (v1 == v2 == 0 which leads to the expection)
p, b = 2.5, 1000: now division by zero due to sc.kv(p, b) does not occur, but nevertheless v1 == v2 == 0 and an error is raised immediately (which is better than the overflow error before)
p, b = 0.75, 1e-6: this works now

I will do a bit more analysis of the histograms to check the part of the method `rvs` that does not rely on `rvs_ratio_uniforms` to be sure that there is no error.

chrisb83 commented:
@WarrenWeckesser I did a bit more testing by checking the histograms of the rvs against the pdf for different parameters and I think the sampling works now. I hope I addressed all your comments above. Could you take a look once more please? Thanks

WarrenWeckesser commented:
@chrisb83, I can probably get back to this next week.  Ping me again if I don't get in another review by next Friday.

WarrenWeckesser commented:
@chrisb83, the examples `geninvgauss.rvs(250, 1.8)` and `geninvgauss.rvs(2.5, 1000)` still cause problems.  Are these fixable?

The former problem with `geninvgauss.rvs(0.75, 1e-6)` is fixed 🎉 , but when I increase `p` I get a warning at `p=1` (when algorithm choice changes):

```
In [180]: geninvgauss.rvs(0.75, 1e-6)                                                                            
Out[180]: 558064.2972103482

In [181]: geninvgauss.rvs(0.999, 1e-6)                                                                           
Out[181]: 2787817.7396335965

In [182]: geninvgauss.rvs(1, 1e-6)                                                                               
/Users/warren/miniconda36/lib/python3.6/site-packages/scipy-1.4.0.dev0+fa97a3a-py3.6-macosx-10.7-x86_64.egg/scipy/stats/_rvs_sampling.py:134: RuntimeWarning: The expected number of iterations to generate a single random number from the desired distribution is larger than 2943033, potentially causing bad performance.
  warnings.warn(msg, RuntimeWarning)
Out[182]: 955122.6659799267
```
I haven't tried to figure out exactly what goes wrong (presumably some floating point loss of precision because of the small value of `b`).  It would be good to see if that can be fixed.

chrisb83 commented:
> the examples geninvgauss.rvs(250, 1.8) and geninvgauss.rvs(2.5, 1000) still cause problems. Are these fixable?

I don't see an easy fix since one cannot evaluate the quasifpdf / pdf in that case:

- if b is large, `exp(-b(x+1/x))` is essentially 0 for all x
- if p is large and p is moderate, the mode is large (~277 if p, b = 250, 1.8). Then if one tries to evaluate the pdf close at the mode, _logquasipdf`is 1151 and exp(...) is inf. You can't evaluate the pdf as well: special.kv(p, b) is inf

chrisb83 commented:
> I get a warning at p=1 (when algorithm choice changes)

True, I understand why now. The algorithm works perfectly fine and it is fast. Looking at the histogram, it fits the pdf closely. 
The problem is that the number of expected iterations is computed as `umax*(vmax - vmin)` and `vmax`and `vmin` normally are computed by evaluating the pdf. If we use the quasipdf, we are using a different scale (which does not matter for the sampling since any scaling factor cancels out if we take the ratio `U/V`). I like the idea of using the quasipdf since does not involve the evaluation of the special function `kv` used in the pdf. But using `rvs_ratio_uniforms` does not work in that case.

I could copy the code of rv_ratio_uniforms and add it here without the check of expected number of iterations (we know this number will be below 2.7 or so). Unfortunately this would duplicate about 40 lines of code. 

Good that you saw this, it could also occur for other parameters.

chrisb83 commented:
> I don't see an easy fix since one cannot evaluate the quasifpdf / pdf in that case

Maybe it can be fixed, I just thought whether one can normalize the pdf in some way that allows for a better evaluation since the scale does not matter. I need to check it.

chrisb83 commented:
Hi Warren, I fixed the sampling for the cases (2.5, 1000) and (250, 1.8) by normalizing the quasidensity. Now we can sample but we cannot evaluate the pdf :) To check whether the samping works, I compared the first and second moments using `mpmath`.

```
np.random.seed(89456)
arglst = [(2.5, 1000), (250, 1.8)]
for args in arglst:
    p, b = args
    r = geninvgauss.rvs(p, b, size=10000)
    mu1 = mpmath.besselk(p + 1, b) / mpmath.besselk(p, b)
    mu2 = mpmath.besselk(p + 2, b) / mpmath.besselk(p, b)
    print(mu1, mu2 - mu1*mu1)
    print(r.mean(), r.var())
```
Results look good:
```
1.003 0.001
1.0033 0.001
277.7814 308.642
278.0164 311.3376
```
It is interesting to see that the results are not "extreme" at all. It would be nice to have an implementation of `special.kv` that can handle large arguments like in `mpmath`.

chrisb83 commented:
I rewrote `_rvs_scalar`: it does not call `rvs_ratio_uniforms` any more. I copied the code of that function and modified it slightly such that it works with the quasidensity (expected number of iterations are not checked to avoid the incorrect warning)

I did additional testing, and all histograms follow the pdf closely:

```
arglist = [(0.3, 0.4), (2.3, 1.5), (0.1, 0.2), (0.9, 1.5), (-0.75, 1),
           (0.8, 0.9), (0, 1.3), (0, 0.9), (0, 0.5)]
for args in arglist:
    r = geninvgauss.rvs(*args, size=10000)
    m = geninvgauss._mode(*args)
    x = np.linspace(0, max(1.5*m, 10), 100)

    fig, ax = plt.subplots(1, 1)
    ax.plot(x, geninvgauss.pdf(x, *args), 'r-', lw=1, alpha=0.6, label='pdf')
    ax.hist(r, bins=50, density=True, histtype='stepfilled', alpha=0.2)
    ax.legend(loc='best', frameon=False) 
    plt.show()
```

chrisb83 commented:
@WarrenWeckesser I hope you have time for another review :) thanks

WarrenWeckesser commented:
@chrisb83, yes, I'll get back to this after this weekend.

chrisb83 commented on scipy/stats/tests/test_distributions.py:
```diff
@@ -312,6 +312,78 @@ def test_pmf(self):
+    @pytest.mark.slow
+    def test_rvs_without_mode_shift(self):
+        # ratio_unif w/o mode shift
+        gig = stats.geninvgauss(0.9, 1.5)
```
This test case needs to be adjusted: current parameters use mode shift (since `b > 1`). I will wait for further feedback before changing it

WarrenWeckesser commented on scipy/stats/_continuous_distns.py:
```diff
@@ -3348,6 +3348,283 @@ def _stats(self, mu):
+    def _logpdf(self, x, p, b):
+        c = -np.log(2*sc.kv(p, b))
+        return _lazywhere(x > 0, (x, p, b, c),
+                          lambda x, p, b, c: c + (p-1)*np.log(x) - b*(x+1/x)/2,
```
PEP 8: Change `p-1` to `p - 1`, and `x+1/x` to `x + 1/x`.  Then wrap the line to keep the line lengths 79 characters or less.

WarrenWeckesser commented on scipy/stats/stats.py:
```diff
@@ -207,7 +207,6 @@
            'combine_pvalues', 'wasserstein_distance', 'energy_distance',
            'brunnermunzel', 'epps_singleton_2samp']
 
-
```
Don't remove this blank line.  From PEP 8: "Surround top-level function and class definitions with two blank lines."

WarrenWeckesser commented:
Another required fix for the docs: add an entry for geninvgauss to doc/source/tutorial/stats/continuous.rst

WarrenWeckesser commented:
It looks like the remaining numerical issue is that the modified Bessel function of the second kind, `kv(v, z)`, overflows to `inf` when `z` is sufficiently close to 0 (where the threshold for "sufficiently close" depends on `v`), and it underflows to 0 when `z` is sufficiently large.

We could eliminate these problems if SciPy had an implementation of the *log* of kv.   For now, we could add checks on the parameter values and raise an error when we know the PDF (and other functions) will produce incorrect results.  For example, we might consider requiring `v < 1000` and `z < 690` (i.e. GIG parameters `p < 1000` and `b < 690`).  That avoids the bad results for "large" `z` by simply not allowing large `z`.

Handling the problems with small `z` is not as clean.  For small `z`, `kv(v, z)` is approximately Γ(v)*(z/2)**(-v)/2 (see, e.g. the last formula in the section ["Asymptotic forms"](https://en.wikipedia.org/wiki/Bessel_function#Asymptotic_forms) of the wikipedia article on Bessel functions, or Abramowitz & Stegun 9.6.9).  Unfortunately, the asymptotic form for small `z` might only give a few digits of precision.  For example,

    In [1281]: def approx_log_kv_small_z(v, z): 
          ...:     return np.log(0.5) + gammaln(v) - v*np.log(z/2) 
          ...:                                                                                        

    In [1282]: v = 400                                                                                

    In [1283]: z = 50.0                                                                               

    In [1284]: float(mpmath.log(mpmath.besselk(v, z)))   # accurate value using mpmath
    Out[1284]: 704.7024066573179

    In [1285]: approx_log_kv_small_z(v, z)                                                            
    Out[1285]: 706.265756308293


We can use the asymptotic form to get a rough boundary of the region in the (`v`, `z`) plane where `kv(v, z)` overflows 64 bit floating point, but computing the result using the log of the asymptotic form in that region means the accuracy of the result depends heavily on the parameter values.  A higher order asymptotic form would be nice, but finding or deriving one might turn into a mini research project. Ping @person142, @pv: any ideas?

pv commented:
For large |z|, maybe https://scipy.github.io/devdocs/generated/scipy.special.kve.html#scipy.special.kve
For small |z|, probably https://dlmf.nist.gov/10.31

WarrenWeckesser commented:
@pv wrote:
> For large |z| (or any |z|), maybe https://scipy.github.io/devdocs/generated/scipy.special.kve.html

Good idea!  We can use that to implement `logkv(v, z)` as `np.log(kve(v, z)) - z`, e.g.

```
In [55]: v = 100                                                                

In [56]: z = 1000                                                               

In [57]: float(mpmath.log(mpmath.besselk(v, z)))                                
Out[57]: -998.2348503607279

In [58]: np.log(kve(v, z)) - z                                                  
Out[58]: -998.2348503607279
```

The small `z` case doesn't look like it will be such a simple one-liner. 😞

WarrenWeckesser commented:
@chrisb83, the above discussion means that with the following two changes, the code should be able to handle much larger values of `b` (e.g. the case `p=2.5`, `b=1000` that came up in earlier reviews):

* In `_logpdf`, change the calculation of `c` to `-np.log(2) - np.log(sc.kve(v, z)) + z`
* Change the calculation of `_munp(n, p, b)` to `np.exp(np.log(sc.kve(p + n, b)) - np.log(sc.kve(p, b)))`

chrisb83 commented:
@WarrenWeckesser nice, i adjusted the code accordingly

WarrenWeckesser commented:
We still have to deal with the incorrect results that can be returned when `kv(v, z)` or `kve(v, z)` overflows to `inf` when `z` is sufficiently small.

As a reminder, here's a specific case where `geninvgauss.pdf` gets it wrong, because `kve(p, b)` is `inf`:

```
In [1]: from scipy.stats import geninvgauss

In [2]: x = 16

In [3]: p = 400

In [4]: b = 50

In [5]: geninvgauss.pdf(x, p, b)
Out[5]: 0.0
```

Here's the result computed using the implementation in `mpsci`:

```
In [6]: from mpsci.distributions import geninvgauss as gig

In [7]: import mpmath

In [8]: mpmath.mp.dps = 25

In [9]: gig.pdf(x, p, b)
Out[9]: mpf('0.4989977639987425652579956449')
```

An implementation of the logarithm of `kv(v, z)` would be useful.

We could, for the moment, accept the limitation in the current implementation by returning `nan` in places where the Bessel function overflows and generate an appropriate warning.  To implement this, it might be as simple as checking the return values of `kv` or `kve` for `inf`, and replacing `inf` with `nan` before proceeding with the rest of the computation.

Or, we hold off on merging this until we have a reasonably accurate implementation of the log of `kv`, or some other way to avoid the overflow problem.

What do folks think of generating a warning and returning `nan` in the overflow case?  For example, we would have something like this:

```
In [33]: geninvgauss.pdf(x, p, [50, 75, 100, 125])
RuntimeWarning: Overflow in internal calculation.  nan returned.
Out[33]: array([            nan, 6.04510713e-017, 5.43158694e-053, 6.33458693e-100])
```
The warning could include more information, if desired.

chrisb83 commented:
Hi Warren,

I would like to add this feature even though we are aware of the limitations:

- the implementation works fine except for quite extreme parameters
- implementation depends on special functions and waiting until logkv is available can take a long time

so i would favor to point out the limitation in the docstring or to replace inf by nan as you proposed.

what do you think?

WarrenWeckesser commented on scipy/stats/_distn_infrastructure.py:
```diff
@@ -40,8 +40,11 @@
     def instancemethod(func, obj, cls):
         return types.MethodType(func, obj)
+
+    from itertools import zip_longest
```
We no longer support Python 2, so this line can be moved out of this `if` block.  That is, we can run `from itertools import zip_longest` unconditionally.  (The clean up of the other conditional import does not have to be done in this pull request.)

WarrenWeckesser commented:
OK, let's go with generating a `RuntimeWarning` and returning `nan` when `kve` overflows to `inf`.  This will require an extra line or to to compute the values of `kve` in an intermediate array and check it for `inf`.  It makes the code a little uglier, and costs some performance, but it prevents us from silently returning nonsensical values.

chrisb83 commented:
Yes, I implemented the warning. Please let me know If you have any suggestions.

WarrenWeckesser commented:
@chrisb83, I'll take another look this weekend.

WarrenWeckesser commented:
OK, the warning should now protect us from silently generating incorrect results:

```
In [83]: geninvgauss.pdf(16, 400, 50)                                                            
/Users/warren/mc37scipy/lib/python3.7/site-packages/scipy-1.4.0.dev0+4b59cd3-py3.7-macosx-10.7-x86_64.egg/scipy/stats/_continuous_distns.py:3511: RuntimeWarning: Infinite values encountered in scipy.special.kve(p, b). Values replaced by NaN to avoid incorrect results.
  warnings.warn(msg, RuntimeWarning)
Out[83]: nan
```

It looks like we need a similar check in `_munp`, to handle a case such as `geninvgauss.mean(240, 10)`. The correct value is approx. 48.0209, but with the current implementation, we get `inf`:

```
In [99]: geninvgauss.mean(240, 10)                                                               
Out[99]: inf
```

chrisb83 commented:
You are right, I added a check to `_munp` (return NaN whenever `kve(p + n, b) / kve(p, b)` is infinite as all moments should be finite)

```
geninvgauss.mean(240, 10)
__main__:340: RuntimeWarning: Infinite value encountered in the moment calculation involving scipy.special.kve. Values replaced by NaN to avoid incorrect results.
Out[21]: nan
```

chrisb83 commented:
@WarrenWeckesser Hi Warren, do you think you can review the changes? I added the warning as you suggested.

WarrenWeckesser commented:
@chrisb83, yes, I'll get to this and your other PR this weekend.

WarrenWeckesser commented on scipy/stats/_continuous_distns.py:
```diff
@@ -3741,7 +3741,15 @@ def _mode(self, p, b):
 
     def _munp(self, n, p, b):
-        return sc.kve(p + n, b) / sc.kve(p, b)
+        m = sc.kve(p + n, b) / sc.kve(p, b)
```
For this to work the way you expect, you'll have to check the numerator and denominator separately.  With the current implementation, with a call such as`geninvgauss.mean(230, 4)`, *both* terms are infinite, so the warning `RuntimeWarning: invalid value encountered in true_divide` is generated by the division and the result is `nan`.  The end result is the same, but instead of the informative warning that you have below, we get a generic warning that doesn't tell the user anything useful.

chrisb83 replied:
True, I changed it now

WarrenWeckesser commented:
Thanks Christoph, looks like the bad cases are covered now.

My personal preference is to squash all the intermediate commits into just a few that contain the essential changes (with appropriate commit messages), and eliminate all the "code churn" that arose during review.  Can you rebase this so that we have just a few essential commits?

chrisb83 commented:
> Can you rebase this so that we have just a few essential commits?

Done

chrisb83 commented:
@WarrenWeckesser All tests are green again after I increased the number of distributions to 100. Can you please check once more? Thanks

WarrenWeckesser commented:
Thanks, Christoph, this is a great addition to SciPy!

WarrenWeckesser commented:
@chrisb83, could you create a PR that mentions this new distribution in the 1.4.0 release notes?

rgommers commented:
> @chrisb83, could you create a PR that mentions this new distribution in the 1.4.0 release notes?

even easier: https://github.com/scipy/scipy/wiki/Release-note-entries-for-SciPy-1.4.0

chrisb83 commented:
> @chrisb83, could you create a PR that mentions this new distribution in the 1.4.0 release notes?
> 
> even easier: https://github.com/scipy/scipy/wiki/Release-note-entries-for-SciPy-1.4.0

Done.

@WarrenWeckesser thanks for your support!
